pkg/server/sotw/v3/ads.go

```diff
@@ -102,55 +81,44 @@ func (s *server) processADS(sw *streamWrapper, reqCh chan *discovery.DiscoveryRe
 				}
 			}
 
-			typeURL := req.GetTypeUrl()
-			subscription, ok := sw.subscriptions[typeURL]
-			if !ok {
-				subscription = stream.NewSubscription(len(req.GetResourceNames()) == 0, nil)
-			}
-
-			if lastResponse, ok := sw.lastDiscoveryResponses[req.GetTypeUrl()]; ok {
-				if lastResponse.nonce == "" || lastResponse.nonce == nonce {
-					// Let's record Resource names that a client has received.
-					subscription.SetReturnedResources(lastResponse.resources)
+			var subscription stream.Subscription
+			w, ok := sw.watches.responders[typeURL]
+			if ok {
+				if w.nonce != "" && req.GetResponseNonce() != w.nonce {
+					// The request does not match the stream nonce, ignore it as per
+					// https://www.envoyproxy.io/docs/envoy/v1.28.0/api-docs/xds_protocol#resource-updates
+					// Ignore this request and wait for the next one
+					// This behavior is being discussed in https://github.com/envoyproxy/envoy/issues/10363
+					// as it might create a race in edge cases, but it matches the current protocol definition
+					s.opts.Logger.Warnf("[sotw ads] Skipping request as nonce is stale for %s", typeURL)
```

Comment:
Curious how often this will trigger in practice. Anytime there are multiple updates sent in sequence or the client is very slow to process updates, we may start seeing this. Info might be enough.

Reply:
Good remark, will set to info

Reply:
Done